design/multiple-listeners-design.md

```diff
@@ -0,0 +1,147 @@
+Root HTTPProxies without TLS enabled have been configured on the HTTP Listener, and those with TLS enabled have been configured on the HTTPS Listener.
+Now that Contour implements Gateway API, it's necessary to support any number of Listeners in order to be fully conformant with the Gateway API spec.
+
+It's important to distinguish between **Envoy Listeners** and **Gateway Listeners**, since they are related but not the same.
+
+An **Envoy Listener** can be thought of simply as Envoy listening on a given port.
+
+A **Gateway Listener** describes listening on a given port, for requests with a given hostname, and optionally with certain TLS details.
+In Envoy terminology, this more closely corresponds to the combination of a Listener, a FilterChain, and a RouteConfiguration virtual host.
+
+## Goals
+- Support Gateways, and associated HTTPRoutes/TLSRoutes, with Listeners configured on >2 ports.
+- Don't break existing functionality for HTTPProxy/Ingress.
+
+## Non Goals
+- Support for HTTPProxy/Ingress attaching to additional Listeners.
+- Support for new route types, e.g. TCPRoute.
+
+## High-Level Design
+Broadly speaking, Contour can operate in one of two modes, which we will call **Gateway mode** and **non-Gateway mode**.
+In Gateway mode, Contour is configured to process a Gateway API Gateway, and gets its Listener configuration from there.
+Non-Gateway mode is the traditional mode of operation, where Listener configuration comes from CLI flags, a config file, and/or a ContourConfiguration resource, plus the Envoy service YAML, and HTTPProxy and/or Ingress is used for routing configuration.
+
+For non-Gateway mode, this design should be fully backwards-compatible.
+There should be no changes observed by a user of Contour in non-Gateway mode.
+
+For Gateway mode, the following changes are needed:
+
+- The Gateway provisioner must configure ports on the Envoy service and daemonset/deployment corresponding to all of the Gateway Listener ports. The mapping between service ports and container ports will be described in the Detailed Design section below.
+- Contour's directed acyclic graph (DAG) processing changes, to first add Listeners, and then add virtual hosts directly to the appropriate Listener(s).
+- Each DAG Listener results in an Envoy Listener being configured. The naming scheme will be described in the Detailed Design section below. 
+- Envoy RouteConfigurations will be generated for each Envoy Listener. The naming scheme will be described in the Detailed Design section below. 
+
+For Gateway mode, we must also consider what happens when HTTPProxy or Ingress is used for routing configuration.
+The details of this use case are described in the Compatibility section below.
+
+## Detailed Design
+
+### internal/gatewayapi
+The `ValidateListeners` function logic must change, to allow any number of unique ports, while still validating protocols and hostnames for each port.
```

Comment:
> allow any number of unique ports

The spec _allows_ for ports on different listeners to be the same:
https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.GatewaySpec

> Each listener in a Gateway must have a unique combination of Hostname, Port, and Protocol.
> 
> An implementation MAY group Listeners by Port and then collapse each group of Listeners into a single Listener if the implementation determines that the Listeners in the group are “compatible”. An implementation MAY also group together and collapse compatible Listeners belonging to different Gateways.
> 
> For example, an implementation might consider Listeners to be compatible with each other if all of the following conditions are met:
> 
> 1. Either each Listener within the group specifies the “HTTP” Protocol or each Listener within the group specifies either the “HTTPS” or “TLS” Protocol.
> 2. Each Listener within the group specifies a Hostname that is unique within the group.
> 3. As a special case, one Listener within a group may omit Hostname, in which case this Listener matches when no other Listener matches.

The language makes it seem that this is not required by implementation, but that it should be done where possible.

For clarity, this example Gateway spec should be supported:
```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
spec:
  listeners:
  - allowedRoutes:
      kinds:
      - group: gateway.networking.k8s.io
        kind: HTTPRoute
    name: https
    hostname: foo.example.com
    port: 443
    protocol: HTTPS
  - allowedRoutes:
      kinds:
      - group: gateway.networking.k8s.io
        kind: TLSRoute
    name: tls
    hostname: bar.example.com
    port: 443
    protocol: TLS
```

Reply:
Agreed, Contour should already allow this for the 2 ports that it does permit, this is covered by this test case: https://github.com/projectcontour/contour/blob/main/internal/gatewayapi/listeners_test.go#L172-L208 (though not 100% sure we've confirmed routing via both an HTTPRoute and TLSRoute works in this scenario).

This line in the doc is indicating that we now need to allow this for >2 ports.